### PR TITLE
[SPARK-45065][PYTHON][PS] Support Pandas 2.1.0

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to support pandas 2.1.0 for PySpark. See [What's new in 2.1.0](https://pandas.pydata.org/docs/dev/whatsnew/v2.1.0.html) for more detail.

### Why are the changes needed?

We should follow the latest version of pandas.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

The existing CI should passed with Pandas 2.1.0

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by itholic** [comment]
Since there are many features are [deprecated from Pandas 2.1.0](https://pandas.pydata.org/docs/whatsnew/v2.1.0.html#deprecations), let me investigate if there is any corresponding feature from Pandas API on Spark while we're here.

File: python/pyspark/pandas/groupby.py
Comment (itholic) [inline python/pyspark/pandas/groupby.py:321]: Bug fixed in Pandas: https://github.com/pandas-dev/pandas/issues/52849.

File: python/pyspark/pandas/tests/test_stats.py
Comment (itholic) [inline python/pyspark/pandas/tests/test_stats.py:282]: The testing util `makeMissingDataframe` is removed.

File: python/pyspark/pandas/typedef/typehints.py
Comment (itholic) [inline python/pyspark/pandas/typedef/typehints.py:490]: Added `dtype` of categories is added to `__repr__`: https://github.com/pandas-dev/pandas/issues/52179.

File: python/pyspark/pandas/tests/frame/test_reshaping.py
Comment (itholic) [inline python/pyspark/pandas/tests/frame/test_reshaping.py]: This just for handling the column order:

**DataFrame**
```python
>>> pdf
    weight height
        kg      m
cat    1.0    2.0
dog    3.0    4.0
```

**DataFrame.stack() in Pandas 1.5.3**
```python
>>> pdf.stack()
        weight  height
cat kg     1.0     NaN
    m      NaN     2.0
dog kg     3.0     NaN
    m      NaN     4.0
```

**DataFrame.stack() in Pandas 2.1.0**
```python
>>> pdf.stack()
        weight  height
cat kg     1.0     NaN
    m      NaN     2.0
dog kg     3.0     NaN
    m      NaN     4.0
```

I think maybe this is the minor bug in Pandas, so I reported for the Pandas community to make sure.

File: python/pyspark/pandas/frame.py
Comment (itholic) [inline python/pyspark/pandas/frame.py:10533]: Bug fixed in Pandas: https://github.com/pandas-dev/pandas/issues/53786.

**Comment by itholic** [comment]
Many tests are failing due to the PyArrow upgrade in CI.

https://github.com/apache/spark/pull/42897 is fixing this issue, so let me rebase the PR after the fixing is get merged.

**Comment by itholic** [comment]
Seems like the CI failure is not related my changes. Let me just try retriggering the CI.

**Comment by itholic** [comment]
I believe this PR is ready to review. The current CI failure is not related to my change.

cc @ueshin @HyukjinKwon @zhengruifeng @xinrong-meng

**Review by HyukjinKwon — APPROVED** [review]

**Comment by zhengruifeng** [comment]
not related to this PR itself, what is the policy to upgrade the minimum version of dependencies listed [here](https://spark.apache.org/docs/latest/api/python/getting_started/install.html#dependencies) ?

**Comment by itholic** [comment]
@zhengruifeng AFAIK, there is no separate policy for minimum version. We may change the minimum version of a particular package when if an older version no longer works properly with Spark, or if the community for that package no longer maintains a particular older version, etc.

**Comment by HyukjinKwon** [comment]
Let's probably upgrade them since we're going ahead for 4.0.0 major version bumpup

**Comment by dongjoon-hyun** [comment]
Could you resolve the conflict, @itholic ?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (Pending CIs)

File: python/pyspark/pandas/frame.py
Comment (ueshin) [inline python/pyspark/pandas/frame.py:1394]: This call will show a deprecation warning from `applymap`?

I guess we should call `return self._apply_series_op(lambda psser: psser.apply(func))` here and `applymap` should call `map` instead?

Comment (itholic) [reply python/pyspark/pandas/frame.py]: Oh, yeah we shouldn't call `applymap` here.

Just applied the suggestion. Thanks!

File: python/docs/source/migration_guide/pyspark_upgrade.rst
Comment (bjornjorgensen) [inline python/docs/source/migration_guide/pyspark_upgrade.rst:44]: Can we add a line her, where we tell users to have pandas version 2.1.0 installed for spark 4.0  
The only way now to find witch pandas version to install is to check the docker file in dev/infra 

https://github.com/jupyter/docker-stacks/blob/52a999a554fe42951e017f7be132d808695a1261/images/pyspark-notebook/Dockerfile#L69

Comment (itholic) [reply python/docs/source/migration_guide/pyspark_upgrade.rst:45]: Good idea. Related information has been added to the top of the migration guide. Thanks!

**Comment by itholic** [comment]
CI link: https://github.com/itholic/spark/actions/runs/6216894150

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM again

The failure `StreamingQueryListenerSuite` is irrelevant to this PR.

Merged to master for Apache Spark 4.0.0.

**Comment by dongjoon-hyun** [comment]
Thank you, @itholic and all!

**Comment by itholic** [comment]
Thanks all!

**Pull request closed without merging**
